### DISCUSSION TITLE
Bulk 301 Redirect

### DISCUSSION THREAD

**Discussion by vimal-here** [vercel/next.js#29063] [Help] [answered]
I have a situation where almost 50,000 (**50K**) old URLs need a `301` permanent redirect. This is to make sure that the old URLs which are indexed by Search Engines, does not throw a 404 when the website is crawled after the next release. 

**The plan:** Create a mapping file (`JSON`) which will read like this

`{"oldUrl_1":"newUrl_1", "oldUrl_2":"newUrl_2", ...};`

I can think of 5 ways by which we can achieve this redirect by **reading** a JSON file.

1. Using a REST client like `axios` 
2. Using require() 
3. Using import()
4. Using `fs.readFile` (Sync) or `fs.readFileAsync` (Async)
5. Using `npm` packages like https://www.npmjs.com/package/readjson

Please suggest the way forward. I am planning to use `fs.readFile`

Note: Any of the above solutions will be called only **once** to move the file contents to a JSON Object (variable/memory) when we do an `npm start`. That means, when a URL hits the browser, no check will be done against the `JSON` file. The URL redirect will happen after checking the presence of old URL in the `JSON` Object (variable).

**Comment by wilsonneto-swe** [comment]
I don't understand... What is the problem here? 

Just put all the URL on the redirects config, you don't more than this....

```
module.exports = {
  async redirects() {
    return [
      {
        source: '/about',
        destination: '/',
        permanent: true,
      },
    ]
  },
}
```

And Next.js will do everything you need...

https://nextjs.org/docs/api-reference/next.config.js/redirects

Let me know if you already know this and this thread is about another problem.

I hope it helps you.
Success in your projects!
Best regards.

**Reply by vimal-here** [reply]
@wilsonneto-dev - Thanks for your reply. 

I understand the inbuilt solution you're proposing. As I mentioned in the first line of my post, we need to consider 50,000 URLs (**50K**).

**Comment by wilsonneto-swe** [comment]
Yes, I understand, but this would be the first test that I would do, put these 50k URLs in the redirects config... Why you think this is a bad way? 50k URLs is a great number for us, humans, but for a code this is small... And I think even if you write your own code and own way to do these redirects, in the end, you will do the same thing next.js will do with the redirects, put all thins on memory then verify if the URL is in array at every request (or generate 50k redirect pages).

I think we don't have an easier way to do this with next.js, another way is to look at what is in front of your next app, a load balancer, a reverse proxy, maybe a good idea to put these redirects there, without using memory and processing power from the next.js app.

I needed something similar with a movies catalog, 8k movies with a lot of backlinks pointing to each, I used:

```js
module.exports = {
  async redirects() {
    return [
      {
        source: '/home',
        destination: '/',
        permanent: true
      },
      {
        source: '/filmes/:slug', // all the movies will redirect
        destination: '/search/:slug', // for example to a "search" or another page
        permanent: true // then the clients will cache this redirect
      },
      // ...
    ];
  }
};
```

I can't imagine 50k URLs without a route pattern like this... You can use it with slugs... Well, I hope it helps you. And sorry if I can't help you more.

Success in your project!
Best regards.

**Reply by vimal-here** [reply]
@wilsonneto-dev - Thanks for giving more clarity. Please correct me if my below understanding is wrong.

> And I think even if you write your own code and own way to do these redirects, in the end, you will do the same thing next.js will do with the redirects, put all thins on memory then verify if the URL is in array at every request

Couple of things here

1. Next config file will go big and, it is no more a configuration file. IMO, it would act as a database like `lowdb`. But if we can move the following to an external file, it will be fine. Not sure if you were also thinking in the same way.
```
      {
        source: '/oldUrl_1',
        destination: '/newUrl_1',
        permanent: true
      },
      {
        source: '/oldUrl_2',
        destination: '/newUrl_2',
        permanent: true
      },
```
2. Next config is a `.js` file, so we will be breaking the rule of thumb of max 500 lines in a JavaScript page. - https://eslint.org/docs/rules/max-lines 
3. Trivial - DRY principle will not be respected. We will have 50K `permanent: true` 
4. Another issue going with Next.js redirect pattern is, we are making the data coupled with Next.js and portability cannot be achieved (but chances to port are minimal). But again, this can be solved by using a JSON file and convert it to an Object (like in point 1) using a custom function.
5. Any solution we use on Next.js side, there will be a performance impact, as all user/browser hits need to verify the URL in our array/Object. At peak load times our application will have ~500 users per second browsing our pages. So one thing we can check here is - Is there a way to bypass this redirect check if the user navigates to this pages inside our application itself. 
Note: The 301 redirect will be applicable only to those pages which Google has indexed. So once a user is inside our application, there are no need for any redirect checks.

something like
```
if (req.headers.referer && req.headers.referer != 'our-domain.xyz') {
    // Verify the array/Object and 301 Redirect
} else {
    // Normal flow
}
```

> another way is to look at what is in front of your next app, a load balancer, a reverse proxy, maybe a good idea to put these redirects there, without using memory and processing power from the next.js app.

Thought of this solution also, but again this has the same issue. All URLs hitting our website will have to go through this rules. Otherwise this is the best way to go about it.

> Success in your project!

With people like you around, world is a better place to code and collaborate. A big thanks for your time.

**Comment by vimal-here** [comment, marked as answer]
Closing this, as we're not going ahead with Bulk redirect as it can be a performance bottleneck for the website.

Thanks @wilsonneto-dev for your time. 👍
